### PR TITLE
Rename *.d.ts to *.d.mts

### LINKED ISSUE #17241: Legacy: Type declarations are broken in 4.0
Ive tested 4.0.189 from NPM and noticed that the type declaration file is broken for legacy with this version. 

> Error:
> Could not find a declaration file for module 'pdfjs-dist/legacy/build/pdf.mjs'

The fix is simple. `pdfjs-dist/legacy/build/pdf.d.ts` should be renamed to `pdfjs-dist/legacy/build/pdf.mjs.d.ts`.

### PR DESCRIPTION
**Pull request by tamuratak**
Rename `*.d.ts` to `*.d.mts`. Close #17241

TypeScript doesn't look up `*.d.ts` files from `*.mjs`. It only looks up `*.d.mts` files.

See 

- https://www.typescriptlang.org/docs/handbook/modules/reference.html#file-extension-substitution

Notice that TypeScript can find `types/src/pdf.d.ts` even though it has the `.d.ts` extension because `"types": "types/src/pdf.d.ts"` is set in `package.json`.

### CODE REVIEW

**Comment by timvandermeij** [comment]
The change itself looks good, but why didn't the typing tests catch this before when we did the change to JavaScript modules? I would assume that `gulp typestest` would have fallen over if the types can't be loaded...

**Comment by tamuratak** [comment]
I have added a test. Is it acceptable that the test depends on `dist`? It is much more reliable to test against the real target.

File: gulpfile.mjs [hunk 1]
```diff
@@ -2344,6 +2312,18 @@ gulp.task(
...
+  gulp.series(setTestEnv, "dist", function runTypesTest(done) {
+    exec('"node_modules/.bin/tsc" -p test/types', function (err, stdout) {
+      if (err) {
+        console.log(`Couldn't compile TypeScript test: ${stdout}`);
+      }
+      done(err);
+    });
+  })
+);
+
```
Comment (Snuffleupagus) [inline gulpfile.mjs:2326]: Why was this code-block moved, since that seems unrelated and just increases the size of the patch?

File: gulpfile.mjs [hunk 2]
```diff
@@ -2344,6 +2312,18 @@ gulp.task(
   })
 );
 
+gulp.task(
+  "typestest",
+  gulp.series(setTestEnv, "dist", function runTypesTest(done) {
```
Comment (Snuffleupagus) [inline gulpfile.mjs:2317]: Do we *need* to run `gulp dist` as part of this, since that basically builds *everything* which will slow down the tests *a lot* just for the typescript definitions?

File: test/types/tsconfig.json [hunk 3]
```diff
@@ -3,7 +3,7 @@
     "outDir": "../../build/tmp",
     "sourceMap": true,
     "declaration": false,
-    "moduleResolution": "node",
+    "moduleResolution": "Node10",
```
Comment (Snuffleupagus) [inline test/types/tsconfig.json:6]: Why this change, since the commit message contains no information about this?

File: test/types/package.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+    "type": "module",
+    "dependencies": {
+        "pdfjs-dist": "../../build/dist/"
+    }
+}
```
Comment (Snuffleupagus) [inline test/types/package.json:6]: This is lacking a new-line at the end of the code, and can we please use *two spaces* of indentation instead?

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: test/types/tsconfig.json (hunk 3 above)
Comment (timvandermeij) [reply test/types/tsconfig.json:6]: Moreover, https://www.typescriptlang.org/tsconfig#moduleResolution indicates we shouldn't rely on this.

**Comment by Snuffleupagus** [comment]
> Is it acceptable that the test depends on `dist`?

As mentioned inline, I don't think so since it'll be slow. Also, as-is it seems that this completely breaks the GitHub Actions tests.

File: gulpfile.mjs (hunk 2 above)
Comment (timvandermeij) [reply gulpfile.mjs:2317]: If we do this I think it needs to be `dist-pre` at least because `dist` is basically `dist-pre` plus committing which fails in the CI currently. I would also like to see the shortest runtime here, but I can also see that testing against the real distribution is perhaps better, so I'm not sure what's best here. If the problem from the issue can be reproduced without `dist-pre` (so basically with the current code) that would be preferable for me too.

Comment (Snuffleupagus) [reply gulpfile.mjs:2317]: If we absolutely must use e.g. `dist-pre` here, then I think that this really should be moved into *its own* workflow separate from the regular GitHub Actions tests and only running in e.g. Node 18.

File: test/types/tsconfig.json (hunk 3 above)
Comment (tamuratak) [reply test/types/tsconfig.json:6]: I have tried that. See https://github.com/mozilla/pdf.js/compare/master...tamuratak:pdf.js:better_types_c

With the `nodenext` option, we have to add the `.js` extension to every `import("./filename")` on JSDoc. The difficult part is that making `tsc-alias` do the job when it rewrites import statements. I had to switch to `tsconfig-replace-paths` instead of `tsc-alias`. 

https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-resolution-is-host-defined

> In Node.js, module specifiers in import statements and dynamic import() calls are not allowed to omit file extensions or /index.js suffixes, while module specifiers in require calls are. This module resolution mode understands and enforces this restriction

**Comment by tamuratak** [comment]
Now the test depends on `generic` as before.

File: test/types/package-lock.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    "requires": true,
+    "packages": {
+        "": {
+            "dependencies": {
+                "pdfjs-dist": "../../build/typestest/"
+            }
+        },
+        "../../build/dist": {
+            "name": "pdfjs-dist",
+            "version": "4.0.201",
```
Comment (Snuffleupagus) [inline test/types/package-lock.json:13]: Here, and elsewhere below, please either use the *correct* version number (via a build-script) or use a dummy value if that's simpler.

Furthermore, how was this file generated and more importantly how will it be updated to adapt to future library changes (this should ideally be automated via a script)?

File: test/types/main.ts
Comment (Snuffleupagus) [inline test/types/main.ts:1]: Should this file now be renamed to `modern.ts` instead, given the new `legacy.ts` added above?

**Comment by Snuffleupagus** [comment]
Are the updated tests running as expected, since I checked out this PR locally and applied the following diff:
```diff
diff --git a/test/types/legacy.ts b/test/types/legacy.ts
index 5a09b683d..48110b216 100644
--- a/test/types/legacy.ts
+++ b/test/types/legacy.ts
@@ -15,6 +15,8 @@ class MainTest {
   }
 }

+throw new Error("legacy");
+
 // This is actually never called, as the test only consists in compiling the file.
 // The compilation will crawl through all files and make sure that the types are consistent.
 const mt = new MainTest("../pdfs/basicapi.pdf");
diff --git a/test/types/modern.ts b/test/types/modern.ts
index 06b892c91..77cd9ffa2 100644
--- a/test/types/modern.ts
+++ b/test/types/modern.ts
@@ -15,6 +15,8 @@ class MainTest {
   }
 }

+throw new Error("modern");
+
 // This is actually never called, as the test only consists in compiling the file.
 // The compilation will crawl through all files and make sure that the types are consistent.
 const mt = new MainTest("../pdfs/basicapi.pdf");
```
before running `gulp typestest` and it actually passed, which suggests that no tests ran!?

File: test/types/tsconfig.json (hunk 3 above)
Comment (timvandermeij) [reply test/types/tsconfig.json]: OK, let's leave that as-is for this patch then.

**Comment by timvandermeij** [comment]
Yes, that should be correct because of the comment below it: type checking happens at compile time, not at run time, so the file itself is not actually run and therefore the exception is not actually thrown.

I have verified that the test do fail correctly if the type annotations are wrong, with the following diff:

```diff
diff --git a/test/types/legacy.ts b/test/types/legacy.ts
index 5a09b683d..7ed6b5805 100644
--- a/test/types/legacy.ts
+++ b/test/types/legacy.ts
@@ -2,7 +2,7 @@ import { getDocument } from "pdfjs-dist/legacy/build/pdf.mjs";
 import { EventBus } from "pdfjs-dist/legacy/web/pdf_viewer.mjs";
 
 class MainTest {
-  eventBus: EventBus;
+  eventBus: boolean;
   task: ReturnType<typeof getDocument> | undefined;
 
   constructor(public file: string) {
diff --git a/test/types/modern.ts b/test/types/modern.ts
index 06b892c91..c0720902f 100644
--- a/test/types/modern.ts
+++ b/test/types/modern.ts
@@ -2,7 +2,7 @@ import { getDocument } from "pdfjs-dist";
 import { EventBus } from "pdfjs-dist/web/pdf_viewer.mjs";
 
 class MainTest {
-  eventBus: EventBus;
+  eventBus: number;
   task: ReturnType<typeof getDocument> | undefined;
 
   constructor(public file: string) {
```

This results in the follow output with a non-zero exit code:

```
Couldn't compile TypeScript test:
test/types/legacy.ts(9,5): error TS2322: Type 'EventBus' is not assignable to type 'boolean'.
test/types/modern.ts(9,5): error TS2322: Type 'EventBus' is not assignable to type 'number'.
```

**Review by timvandermeij — APPROVED** [review]
Looks good to me, but please squash the commits before we can merge it. Thank you for fixing this and improving the tests!

**Review by Snuffleupagus — APPROVED** [review]

**Comment by timvandermeij** [comment]
Thanks!
